src/averageAges.js

```diff
@@ -52,7 +66,43 @@ function calculateWomenAverageAge(people, withChildren) {
  * @return {number}
  */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // write code here
+  const mothersNamesArray
+  = people.map((child) => child.mother);
+
+  const mInfo
+  = people.filter(mother => mothersNamesArray.includes(mother.name)
+  && mother.name !== null);
+
+  const peopleWithMothersInList
+  = people.filter(child => mInfo.find(mother => mother.name === child.mother));
+
+  const validPeopleWithMothersInList
+  = peopleWithMothersInList.filter(child => child.mother !== null);
+
+  const fullInfoValidPeopleWithMothersInList
+  = validPeopleWithMothersInList.map(child => {
+    child.mothersBirthYear
+    = mInfo.find(mother => mother.name === child.mother).born;
+
+    return child;
+  });
+
+  const fullInfoValidMansWithMothersInList
+  = fullInfoValidPeopleWithMothersInList.filter(child => child.sex === 'm');
+
+  let diferences;
+
+  onlyWithSon === undefined
+    ? diferences
+    = fullInfoValidPeopleWithMothersInList.map(child => child.born
+      - child.mothersBirthYear)
+    : diferences
+    = fullInfoValidMansWithMothersInList.map(child => child.born
+      - child.mothersBirthYear);
+
+  const sum = diferences.reduce((s, difference) => s + difference, 0);
+
+  return sum / diferences.length;
```

Comment:
that's too many strings of code, it should be shorter, I'll give you a hint:
split this task into blocks and implement them one by one:
1) create an array with kids who have a mother (use ternary operator for deciding which condition choose don't repeat your code)
2) create an array with a difference between a child "born" field and mother "born" field, which equal to the age of the mother when she born her child 
3)and then just find the average age from the previous array